src/main/java/com/iota/iri/network/NeighborRouter.java

```diff
@@ -0,0 +1,866 @@
+                * Converter.NUMBER_OF_TRITS_IN_A_TRYTE];
+        Converter.trits(config.getCoordinator().toString(), tritsEncodedCooAddress, 0);
+        byteEncodedCooAddress = new byte[Protocol.BYTE_ENCODED_COO_ADDRESS_BYTES_LENGTH];
+        Converter.bytes(tritsEncodedCooAddress, byteEncodedCooAddress);
+    }
+
+    private void initNeighbors() {
+        // parse URIs
+        config.getNeighbors().stream()
+                .distinct()
+                .map(NeighborRouter::parseURI)
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .forEach(uri -> reconnectPool.add(uri));
+    }
+
+    /**
+     * Starts a dedicated thread for the {@link NeighborRouter} and then starts the routing mechanism.
+     */
+    public void start() {
+        executor.execute(this::route);
+    }
+
+    /**
+     * Starts the routing mechanism which first initialises connections to neighbors from the configuration and then
+     * continuously reads and writes messages from/to neighbors. <br/>
+     * This method will also try to reconnect to wanted neighbors by the given
+     * {@link BaseIotaConfig#getReconnectAttemptIntervalSeconds()} value.
+     */
+    public void route() {
+        log.info("starting neighbor router");
+
+        // run selector loop
+        try {
+            selector = Selector.open();
+            serverSocketChannel = ServerSocketChannel.open();
+            serverSocketChannel.configureBlocking(false);
+            serverSocketChannel.register(selector, SelectionKey.OP_ACCEPT);
+            InetSocketAddress tcpBindAddr = new InetSocketAddress(config.getNeighboringSocketAddress(),
+                    config.getNeighboringSocketPort());
+            serverSocketChannel.socket().bind(tcpBindAddr);
+            log.info("bound server TCP socket to {}", tcpBindAddr);
+
+            // parse neighbors from configuration
+            initNeighbors();
+
+            // init connections to the wanted neighbors,
+            // this also ensures that the whitelists are updated with the corresponding
+            // IP addresses and domain name mappings.
+            connectToWantedNeighbors();
+
+            long lastReconnectAttempts = System.currentTimeMillis();
+            long reconnectAttemptTimeout = TimeUnit.SECONDS.toMillis(config.getReconnectAttemptIntervalSeconds());
+
+            while (!shutdown.get()) {
+                selector.select(reconnectAttemptTimeout);
+                if (shutdown.get()) {
+                    break;
+                }
+
+                // reinitialize connections to wanted neighbors
+                long now = System.currentTimeMillis();
+                if (forceReconnectAttempt.get() || now - lastReconnectAttempts > reconnectAttemptTimeout) {
+                    lastReconnectAttempts = now;
+                    forceReconnectAttempt.set(false);
+                    connectToWantedNeighbors();
```

Comment:
I don't think that initiated connections above the same loop before select() will have interests already available on the next select() call. I will leave it like this for now because even if it did, it would really only decrease latency in an insignificant way.